README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# arabic2kansuji

<!-- badges: start -->
<!-- badges: end -->

`{arabic2kansuji}` is a package consisting of Simple functions to
convert given Arabic numerals to “Kansuji” numerical figures that
represent numbers written in Chinese characters.

## Installation

You can install the released version of `{arabic2kansuji}` from CRAN
with:

``` r
install.packages("arabic2kansuji")
```

and also, install the development version install from GitHub:

``` r
install.packages("remotes")
remotes::install_github("indenkun/arabic2kansuji")
```

## Example

load library.

``` r
library(arabic2kansuji)
```

### `arabic2kansuji`

`arabic2kansuji` is a function that converts Arabic numerals to Kansuji
verbatim.

There is no need to enclose only half-width Arabic numerals with
double-quotation marks.

``` r
arabic2kansuji(1234567890)
#> [1] "一二三四五六七八九〇"
```

If you give a string containing Arabic numerals, everything else is
retained except the Arabic numerals to be converted. Strings containing
Arabic numerals can be converted, but please enclose them in
double-quotations

``` r
arabic2kansuji("昭和64年は1989年1月7日までです。")
#> [1] "昭和六四年は一九八九年一月七日までです。"
```

By default, it converts only half-width Arabic numerals, but you can
also convert only full-width Arabic numerals or convert both full-width
and half-width Arabic numerals to Kansuji by specifying arguments.

``` r
# By default, full-width Arabic numerals will not be converted.
arabic2kansuji("東京都新宿区西新宿２丁目８−１")
#> [1] "東京都新宿区西新宿２丁目８−１"
# It can convert full-width Arabic numerals by providing an argument.
arabic2kansuji("東京都新宿区西新宿２丁目８−１", width = "full")
#> [1] "東京都新宿区西新宿二丁目八−一"
arabic2kansuji("全角アラビア数字１２３と半角アラビア数字123の混在も引数を指定すると変換できます。", width = "all")
#> [1] "全角アラビア数字一二三と半角アラビア数字一二三の混在も引数を指定すると変換できます。"
```

By default, 0 is converted to rei (〇), but it can be converted to zero
(零) by specifying an argument.

``` r
arabic2kansuji("令和２年は2020年")
#> [1] "令和２年は二〇二〇年"
arabic2kansuji("令和2年は2020年", zero = "zero")
#> [1] "令和二年は二零二零年"
```

Since this function only replaces Arabic numerals with Kansuji verbatim,
it cannot be converted by calculating 1234 as 千二百三十四. This feature
is provided by the `arabic2kansuji_num` functions.

### `arabic2kansuji_num`

This `arabic2kansuji_num` calculates and converts a given number to a
kanji value. For example, 1234 can be converted to 千二百三十四.

``` r
arabic2kansuji_num(124271318)
#> [1] "一億二千四百二十七万千三百十八"
```

However, currently only accept one half-width Arabic numerals.

Use `arabic2kansuji_num` if you want to convert two or more half-width
Arabic numerals by calculating the number of Chinese characters. You can
also use `arabic2kansuji_all` to convert full-width Arabic numerals (as
a string) to Kansuji.

``` r
arabic2kansuji_num("124271318人")
#> Warning in .f(.x[[i]], ...): only number can convert to kansuji.
#> [1] NA
```

Use `arabic2kansuji_all` to calculate and convert Arabic numerals to
Kansuji while keeping the string in Arabic numerals containing the
string.

For more than 17 digits, a warning is given because calculation
processing may not be performed correctly due to problems on the R.

``` r
arabic2kansuji_num(1234567890123456789)
#> Warning in .f(.x[[i]], ...): too long number to convert exactly.
#> [1] "百二十三京四千五百六十七兆八千九百一億二千三百四十五万六千七百六十八"
```

Negative values are also supported. The notation will be 負Kansuji.

``` r
arabic2kansuji_num(-123456789)
#> [1] "負一億二千三百四十五万六千七百八十九"
```

Too large values (more than 10 to the power of 20) cannot be calculated.

``` r
arabic2kansuji_num(100000000000000000000)
#> Warning in .f(.x[[i]], ...): too large number to convert.
#> [1] NA
```

`arabic2kansuji_num` accepts two or more Arabic numerals, calculates and
converts them to Kansuji.

``` r
x <- c(123, 456, 789)
arabic2kansuji_num(x)
#> [1] "百二十三"   "四百五十六" "七百八十九"
```

### `arabic2kansuji_all`

`arabic2kansuji_num` only accepts half-width Arabic numbers, but
`arabic2kansuji_all` can convert a string containing a half-width Arabic
number to Kansuji and keep the string. `arabic2kansuji_all` can convert
not only half-width Arabic numerals but also full-width Arabic numerals
can be converted by specifying an argument. However, it is not possible
to convert only the full-size Arabic numerals in the string and keep the
half-size Arabic numerals.

``` r
arabic2kansuji_all("昭和64年は1989年1月7日までです。")
#> [1] "昭和六十四年は千九百八十九年一月七日までです。"
arabic2kansuji_all("平成３１年は2019年4月30日までです。")
#> [1] "平成３１年は二千十九年四月三十日までです。"
arabic2kansuji_all("平成３１年は2019年4月30日までです。", widths = "all")
#> [1] "平成三十一年は二千十九年四月三十日までです。"
```

`arabic2kansuji_all` can receive more than one string.

``` r
x <- c("昭和64年は1989年1月7日までです。", "平成31年は2019年4月30日までです。")
arabic2kansuji_all(x)
#> [1] "昭和六十四年は千九百八十九年一月七日までです。"
#> [2] "平成三十一年は二千十九年四月三十日までです。"
```

`arabic2kansuji_all` can also convert Arabic numerals and Kansuji
intended for digits to represent a single number, such as 1億2345万
intended for 一億二千三百四十五万. This is a kind of side effect that
happens when Arabic numbers are converted to Kansuji because the
readings are just the same.

``` r
arabic2kansuji_all("1億2345万")
#> [1] "一億二千三百四十五万"
```

## Imports packages

-   `{purrr}`
-   `{stringr}`
-   `{stats}`

## Known Issue

-   `arabic2kansuji_all` a combination of Arabic and Kansuji
    representing a single number of digits, such as tens or billions,
    but not including the next upper tens or billions of digits, such as
    12345万, which is intended to be 一億二千三百四十五万, would be
    valued at 一万二千三百四十五 in the Arabic numeral place and would
    not appear in the intended form. There is currently no error on this
    issue.

``` r
arabic2kansuji_all("12345万")
#> [1] "一万二千三百四十五万"
```

-   `arabic2kansuji_*` does not support conversion of values greater
    than one hundred quintillion. Entering a value greater than one
    hundred quintillion will return a warning message and NA.

``` r
arabic2kansuji_num(1e+20)
#> Warning in .f(.x[[i]], ...): too large number to convert.
#> [1] NA
arabic2kansuji_all("100000000000000000000個")
#> Warning in .f(.x[[i]], ...): too large number to convert.
#> [1] NA
```

## License

MIT.


NEWS.md
# arabic2kansuji (development version)

# arabic2kansuji 0.1.3

* Made it possible to correct conversions of some numbers to Chinese numerals that are not erroneous but are uncomfortable for native speakers.
    * A new argument (`add.one_thousand` and `add.one_thousand.all`) has been added to specify whether or not 1 should be displayed as a Chinese numeral when the thousandth digit is 1.

# arabic2kansuji 0.1.2

* Fixed the situation so that the NOTE of the check does not appear in CRAN.

# arabic2kansuji 0.1.1

* FIX BUG
    * Fixed a bug that prevented the return of a value when a large value such as 1e23 was entered, and changed it to return NA with an error message.
 
# arabic2kansuji 0.1.0

* Initial CRAN release.


R/arabic2kansuji.R
#' Convert Arabic numerals to Kansuji
#' @description
#' Converts a given Arabic numerals to Kansuji numerical figures that written
#' in Chinese characters (in other words, Chinese numeral). \code{arabic2kansuji()}
#' function can also convert Arabic numerals in the string to kansuji.
#' \code{arabic2kansuji_num()} function can convert a vector of only one Arabic
#' numerals to Kansuji. Use \code{arabic2kansuji_all()} can calculate and convert
#' Arabic numerals to Kansuji while retaining the original non-Arabic numeral
#' string.
#' @param str Input vector.
#' @param zero Selects whether 0 should be zero like o ("rei") or zero like have
#'             rain crown ("zero") when Arabic numerals are converted to kansuji.
#' @param width Selects whether you want to convert Arabic numbers, only
#'              half-width numbers ("half") or only full-width
#'              numbers ("full") or both half-width and full-width
#'              numbers ("all") when converting Arabic numbers to kansuji.
#'
#' @rdname arabic2kansuji
#' @return a character.
#' @examples
#' arabic2kansuji(2020)
#' arabic2kansuji(2020, zero = "zero")
#' arabic2kansuji(c(2019, 2020, 2021))
#' arabic2kansuji_num(2020)
#' arabic2kansuji_num((c(2019, 2020, 2021)))
#' arabic2kansuji_all("This year is 2020, next year is 2021.")
#' @export
#'
arabic2kansuji <- function(str,
                           zero = c("rei", "zero"),
                           width = c("half", "full", "all")){

  zero <- match.arg(zero)
  width <- match.arg(width)

  arabicn_half <- "1234567890"
  arabicn_full <- "\uff11\uff12\uff13\uff14\uff15\uff16\uff17\uff18\uff19\uff10"

  if(zero == "rei") kansuji <- "\u4e00\u4e8c\u4e09\u56db\u4e94\u516d\u4e03\u516b\u4e5d\u3007"
  else if(zero == "zero") kansuji <- "\u4e00\u4e8c\u4e09\u56db\u4e94\u516d\u4e03\u516b\u4e5d\u96f6"

  arabicn_half <- unlist(stringr::str_split(arabicn_half, ""))
  arabicn_full <- unlist(stringr::str_split(arabicn_full, ""))
  kansuji <- unlist(stringr::str_split(kansuji, ""))

  if(width == "half") arabicn <- arabicn_half
  else if(width == "full") arabicn <- arabicn_full
  else if(width == "all"){
    arabicn <- c(arabicn_half, arabicn_full)
    kansuji <- c(kansuji, kansuji)
  }

  names(kansuji) <- arabicn

  stringr::str_replace_all(str, kansuji)

}

#'
arabic2kansuji_cal <- function(num, add.one_thousand = TRUE, add.one_thousand.all = FALSE, ...){
  if(length(num) > 1) stop("only one number can convert to kansuji. use `arabic2kansuji_cal` to convert to over 2 numbers.")
  if(!is.numeric(num)){
    warning("only number can convert to kansuji.")
    return(NA)
  }
  if(is.na(num) || is.nan(num) || is.infinite(num) || is.null(num)){
    warning("only number can convert to kansuji.")
    return(NA)
  }

  negative <- 0
  if(num < 0){
    negative <- 1
    num <- abs(num)
  }

  num.str <- prettyNum(num, scientific = FALSE)

  n <- purrr::reduce(stringr::str_split(num.str,
                                        stringr::boundary(type = "character")),
                     c)
  n <- suppressWarnings(as.numeric(n))

  if(length(stats::na.omit(n)) == 1 && stats::na.omit(n) == 0) return(arabic2kansuji(stats::na.omit(n), ...))

  if(any(num >= 1e+17 && num < 1e+20)){
    warning("too long number to convert exactly.")
  }else if(num >= 1e+20){
    warning("too large number to convert.")
    return(NA)
  }

  m <- max(which(n >= 0)) - which(n >= 0) + 1
  names(n) <- m

  for(i in 1:length(n)){
    if(as.numeric(names(n[i])) %% 4 == 1){
      if(n[i] >= 1) n[i] <- arabic2kansuji(n[i])
    }
    else if(as.numeric(names(n[i])) %% 4 == 2){
      if(n[i] >= 2) n[i] <- paste0(arabic2kansuji(n[i]), "\u5341")
      else if(n[i] == 1) n[i] <- "\u5341"
    }
    else if(as.numeric(names(n[i])) %% 4 == 3){
      if(n[i] >= 2) n[i] <- paste0(arabic2kansuji(n[i]), "\u767e")
      else if(n[i] == 1) n[i] <- "\u767e"
    }
    else if(as.numeric(names(n[i])) %% 4 == 0){
      if(add.one_thousand){
        if(add.one_thousand.all){
          if(n[i] != 0) n[i] <- paste0(arabic2kansuji(n[i]), "\u5343")
        }else{
          if((as.numeric(names(n[i])) %/% 4) != 1 && n[i] != 0) n[i] <- paste0(arabic2kansuji(n[i]), "\u5343")
          else if((as.numeric(names(n[i])) %/% 4) == 1 && n[i] >= 2) n[i] <- paste0(arabic2kansuji(n[i]), "\u5343")
          else if((as.numeric(names(n[i])) %/% 4) == 1 && n[i] == 1) n[i] <- "\u5343"
        }
      }else if(n[i] >= 2) n[i] <- paste0(arabic2kansuji(n[i]), "\u5343")
      else if(n[i] == 1) n[i] <- "\u5343"
    }
  }

  if(length(n) >= 5){
    if(length(n) >= 9) l <- length(n) - 7
    else l <- 1
    for(i in (length(n) - 4):l){
      if(n[i] != 0){
        n[i] <- paste0(n[i], "\u4e07")
        break
      }
    }
  }
  if(length(n) >= 9){
    if(length(n) >= 13) l <- length(n) - 11
    else l <- 1
    for(i in (length(n) - 8):l){
      if(n[i] != 0){
        n[i] <- paste0(n[i], "\u5104")
        break
      }
    }
  }
  if(length(n) >= 13){
    if(length(n) >= 17) l <- length(n) - 15
    else l <- 1
    for(i in (length(n) - 12):l){
      if(n[i] != 0){
        n[i] <- paste0(n[i], "\u5146")
        break
      }
    }
  }
  if(length(n) >= 17){
    if(length(n) >= 21) l <- length(n) - 19
    else l <- 1
    for(i in (length(n) - 16):l){
      if(n[i] != 0){
        n[i] <- paste0(n[i], "\u4eac")
        break
      }
    }
  }

  n <- gsub("0", NA, n)
  if(negative == 1) n <- c("\u8ca0", n)
  stringr::str_flatten(stats::na.omit(n))

}

#' @param num Input only number. Accept more than one number.
#' @param add.one_thousand a logical whether or not 1 should be displayed as a
#'                         Kansuji numeral when the thousandth digit is 1.
#' @param add.one_thousand.all Specifies whether or not to display the Kansuji
#'                             numeral 1 when the number is 1,000. This is valid
#'                             only when \code{add.one_thousand} is TRUE.
#' @rdname arabic2kansuji
#' @export
#'
arabic2kansuji_num <- function(num, add.one_thousand = TRUE, add.one_thousand.all = FALSE, ...){

  purrr::map2_chr(num, add.one_thousand, arabic2kansuji_cal, add.one_thousand.all, ...)
}

#'
arabic2kansuji_all_num <- function(str, widths = c("half", "all"), add.one_thousand = TRUE, add.one_thousand.all = FALSE, ...){
  widths <- match.arg(widths)

  str_row <- str
  if(widths == "all"){
    arabicn_half <- "1234567890"
    arabicn_full <- "\uff11\uff12\uff13\uff14\uff15\uff16\uff17\uff18\uff19\uff10"

    arabicn_half <- unlist(stringr::str_split(arabicn_half, ""))
    arabicn_full <- unlist(stringr::str_split(arabicn_full, ""))

    names(arabicn_half) <- arabicn_full

    str <- stringr::str_replace_all(str, arabicn_half)
  }

  str_num <- stringr::str_split(str, pattern = "[^0123456789]")[[1]]
  str_num[str_num == ""] <- NA
  str_num <- stats::na.omit(str_num)
  str_kansuji <- arabic2kansuji_num(as.numeric(str_num), add.one_thousand, add.one_thousand.all, ...)

  if(length(str_num) == 0) return(str_row)

  for(i in 1:length(str_num)){
    str <- stringr::str_replace(str, pattern = str_num[i], replacement = str_kansuji[i])
  }

  return(str)
}

#' @param widths Selects whether you want to convert Arabic numbers, only
#'               half-width numbers ("half") or both half-width and
#'               full-width numbers ("all") when converting Arabic numbers to
#'               Kansuji.
#' @param ...    Other arguments to carry over to `arabic2kansuji()`.
#' @rdname arabic2kansuji
#' @export
#'
arabic2kansuji_all <- function(str, widths = c("half", "all"), add.one_thousand = TRUE, add.one_thousand.all = FALSE, ...){
  widths <- match.arg(widths)
  purrr::map2_chr(str, widths, arabic2kansuji_all_num, add.one_thousand, add.one_thousand.all, ...)
}
